cpp/ql/src/Diagnostics/FailedExtractions.qll

```diff
@@ -0,0 +1,121 @@
+import cpp
+
+/**
+ * The class of errors upon we mark a file as non-successfully extracted.
+ */
+class ReportableError extends Diagnostic {
+  ReportableError() {
+    (
+      this instanceof CompilerDiscretionaryError or
+      this instanceof CompilerError or
+      this instanceof CompilerCatastrophe
+    ) and
+    // If the extractor encounters an error in a compilation, it always emits a
+    // catch-all diagnostic "There was an error during this compilation", to ensure
+    // that the error makes it to the database.
+    // This error doesn't have a file path attached to it, and is thus
+    // useless for us to report. Furthermore, in the common case, we will have a
+    // proper diagnostic for this error we can show.
+    // Instead, we synthesize `TUnknownError` if this is the only error that we can show to the user.
+    not this.getFile().getAbsolutePath() = ""
+  }
+}
+
+private newtype TExtractionError =
+  TReportableError(ReportableError err) or
+  TCompilationFailed(Compilation c, File f) {
+    f = c.getAFileCompiled() and not c.normalTermination()
+  } or
+  // Report generic extractor errors only if we haven't seen any other error-level diagnostic
+  TUnknownError(CompilerError err) { not exists(ReportableError e) }
+
+/**
+ * Superclass for the extraction error hierarchy.
+ */
+class ExtractionError extends TExtractionError {
+  string toString() { none() }
+
+  /** Gets the error message for this error. */
+  string getErrorMessage() { none() }
+
+  /** Gets the file this error occured in. */
+  File getFile() { none() }
+
+  /** Gets the location this error occured in. */
+  Location getLocation() { none() }
+
+  /** Gets the SARIF severity of this error. */
+  int getSeverity() {
+    // Unfortunately, we can't distinguish between errors and fatal errors in SARIF,
+    // so all errors have severity 2.
+    result = 2
+  }
+}
+
+/**
+ * An irrecoverable extraction failure, where extraction was unable to finish.
+ * This can be caused by a multitude of reasons, for example:
+ *  - hitting an EDG assertion
+ *  - crashing due to dereferencing an invalid pointer
+ *  - stack overflow
+ *  - out of memory
+ */
+class ExtractionIrrecoverableError extends ExtractionError, TCompilationFailed {
+  Compilation c;
+  File f;
+
+  ExtractionIrrecoverableError() { this = TCompilationFailed(c, f) }
+
+  override string toString() {
+    result = "Irrecoverable extraction error while compiling " + f.toString()
+  }
+
+  override string getErrorMessage() {
+    result =
+      "Irrecoverable compilation failure, check logs/build-tracer.log in the database directory for more information."
```

Comment:
This should use a semicolon, not a comma. Also, I think the log isn't kept by default, so perhaps we can link to a page giving debugging instructions instead?

Reply:
The log is kept if you invoke `codeql database create`, LGTM removes it from the databases offered there, though.
If we have such a page, I'm happy to link to it, but I don't know of any...
I don't know about the behaviour of the action, but if you enable the step debug, the actions prints it even to the actions log (I just learned about that yesterday, very exciting!).

Reply:
> The log is kept if you invoke codeql database create

Right, but we expect this to mostly be used with code-scanning, which doesn't keep it by default.

Reply:
As I learned yesterday, it's even worse - the actual log is not accessible by the user anywhere.
Maybe I should replace the message with "Irrecoverable compilation failure, please contact support if the problem persists."?
@jbj do you have an opinion?

Reply:
I prefer keeping organisational matters like "contact support" out of the code. Just "Irrecoverable compilation failure" sounds fine.

Reply:
That's a good argument, I agree.